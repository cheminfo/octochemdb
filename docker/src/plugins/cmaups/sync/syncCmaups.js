import debugLibrary from '../../../utils/Debug.js';
import { getTaxonomiesForCmaupsAndNpasses } from '../../activesOrNaturals/utils/utilsTaxonomies/getTaxonomiesForCmaupsAndNpasses.js';
import { taxonomySynonyms } from '../../activesOrNaturals/utils/utilsTaxonomies/taxonomySynonyms.js';

import cmaupsStartSync from './utils/cmaupsStartSync.js';
import { getNormalizedActivities } from './utils/getNormalizedActivities.js';
import { parseCmaups } from './utils/parseCmaups.js';

const debug = debugLibrary('syncCmaups');
/**
 * @description sync the cmaups collection
 * @param {*} connection the connection to the database
 * @return {Promise} returns cmaups collection
 */
export async function sync(connection) {
  try {
    // Read files to be parsed, get last document imported, progress, sources and logs
    const [
      lastDocumentImported,
      progress,
      sources,
      collection,
      general,
      activities,
      speciesPair,
      speciesInfo,
      logs,
    ] = await cmaupsStartSync(connection);
    // Define counters
    let counter = 0;
    let imported = 0;
    let start = Date.now();
    let isTimeToUpdate = false;
    if (
      progress.dateEnd !== 0 &&
      Date.now() - progress.dateEnd >
        Number(process.env.CMAUP_UPDATE_INTERVAL) * 24 * 60 * 60 * 1000 &&
      JSON.stringify(sources) !== progress.sources
    ) {
      progress.dateStart = Date.now();
      await connection.setProgress(progress);
      isTimeToUpdate = true;
    }
    // Reimport collection again only if lastDocument imported changed or importation was not completed
    if (
      lastDocumentImported === null ||
      ((JSON.stringify(sources) !== progress.sources ||
        progress.state !== 'updated') &&
        isTimeToUpdate)
    ) {
      // get old to new taxonomies ids and taxonomies collection
      const oldToNewTaxIDs = await taxonomySynonyms();
      const collectionTaxonomies = await connection.getCollection('taxonomies');
      // Define stat updating because in case of failure Cron will retry importation in 24h
      progress.state = 'updating';
      await connection.setProgress(progress);
      // Create a temporaty collection to avoid to drop the data already imported before the new ones are ready
      const temporaryCollection = await connection.getCollection('cmaups_tmp');
      debug(`Start parsing cmaup`);
      for await (const entry of parseCmaups(
        general,
        activities,
        speciesPair,
        speciesInfo,
        connection,
      )) {
        counter++;
        // If cron launched in mode test, the importation will be stopped after 20 iteration
        if (process.env.NODE_ENV === 'test' && counter > 20) break;

        // Debug the processing progress every 10s or the defined time in process env
        if (Date.now() - start > Number(process.env.DEBUG_THROTTLING)) {
          debug(`Processing: counter: ${counter} - imported: ${imported}`);
          start = Date.now();
        }
        /// Normalize Taxonomies
        if (entry.data.taxonomies) {
          let taxonomies = await getTaxonomiesForCmaupsAndNpasses(
            entry,
            collectionTaxonomies,
            oldToNewTaxIDs,
            'cmaups',
          );
          entry.data.taxonomies = taxonomies;
        }
        // Normalize activities
        if (entry.data.activities) {
          let activities = await getNormalizedActivities(
            entry,
            collectionTaxonomies,
            oldToNewTaxIDs,
            'cmaups',
          );
          entry.data.activities = activities;
        }
        // Insert the entry(i) in the temporary collection
        entry._seq = ++progress.seq;
        await temporaryCollection.updateOne(
          { _id: entry._id },
          { $set: entry },
          { upsert: true },
        );
        imported++;
      }
      // Once it is finished, the temporary collection replace the old collection
      await temporaryCollection.rename('cmaups', {
        dropTarget: true,
      });
      // Define logs informations in collection importationLogs
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'updated';
      await connection.updateImportationLog(logs);
      // Define new informations and set state to updated in admin collection
      progress.sources = sources;
      progress.dateEnd = Date.now();
      progress.state = 'updated';
      await connection.setProgress(progress);
      // Indexing of properties in collection
      await collection.createIndex({ 'data.ocl.noStereoTautomerID': 1 });
      await collection.createIndex({ _seq: 1 });

      debug(`${imported} compounds processed`);
    } else {
      debug(`file already processed`);
    }
  } catch (e) {
    // If error is catch, debug it on telegram
    if (connection) {
      debug(e.message, { collection: 'cmaups', connection, stack: e.stack });
    }
  }
}
