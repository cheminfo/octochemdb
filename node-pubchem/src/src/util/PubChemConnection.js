'use strict';

const delay = require('delay');
const { MongoClient } = require('mongodb');
const debug = require('debug')('PubChemConnection');

function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function getProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.updateOne({ _id: progress._id }, { $set: progress });
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    debug('Starting new database construction.');
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  } else {
    if (progress.state === 'update') {
      debug('First importation has been completed. Should only update.');
    } else {
      debug(`Continuing first importation from ${progress.seq}.`);
    }
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      console.log(e);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};

module.exports = PubChemConnection;
