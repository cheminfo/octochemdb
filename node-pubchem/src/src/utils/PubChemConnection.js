import Debug from 'debug';
import delay from 'delay';
import { MongoClient } from 'mongodb';

const debug = Debug('PubChemConnection');

export function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function setProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      console.log(e);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
