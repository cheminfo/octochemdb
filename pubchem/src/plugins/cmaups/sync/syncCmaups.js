import Debug from '../../../utils/Debug.js';

import { parseCmaups } from './utils/parseCmaups.js';
import cmaupsStartSync from './utils/cmaupsStartSync.js';

const debug = Debug('syncCmaups');

export async function sync(connection) {
  try {
    const {
      firstID,
      lastDocumentImported,
      progress,
      sources,
      collection,
      general,
      activities,
      speciesPair,
      speciesInfo,
      logs,
    } = await cmaupsStartSync(connection);
    await collection.createIndex({ ' _seq': 1 });
    let skipping = firstID !== undefined;
    let counter = 0;
    let imported = 0;
    let start = Date.now();

    // Reimport collection again only if lastDocument imported changed or importation was not completed
    if (
      lastDocumentImported === null ||
      sources !== progress.sources ||
      progress.state !== 'updated'
    ) {
      let parseSkip;
      if (skipping && progress.state !== 'updated') {
        parseSkip = firstID;
      }
      debug(`Start parsing cmaup`);
      for await (const entry of parseCmaups(
        general,
        activities,
        speciesPair,
        speciesInfo,
        parseSkip,
        connection,
      )) {
        counter++;
        // If test mode break process after counter >20
        if (process.env.TEST === 'true' && counter > 20) break;

        // Log dubug each 10s or defined time
        if (
          Date.now() - start >
          Number(process.env.DEBUG_THROTTLING || 10000)
        ) {
          debug(`Processing: counter: ${counter} - imported: ${imported}`);
          start = Date.now();
        }

        entry._seq = ++progress.seq;
        progress.state = 'updating';
        await collection.updateOne(
          { _id: entry._id },
          { $set: entry },
          { upsert: true },
        );
        await connection.setProgress(progress);
        imported++;
      }
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'updated';
      await connection.updateImportationLog(logs);
      progress.sources = sources;
      progress.date = new Date();
      progress.state = 'updated';
      await connection.setProgress(progress);
      debug(`${imported} compounds processed`);
    } else {
      debug(`file already processed`);
    }
    // we remove all the entries that are not imported by the last file
    const result = await collection.deleteMany({
      _seq: { $lte: logs.startSequenceID },
    });
    debug(`Deleting entries with wrong source: ${result.deletedCount}`);
  } catch (e) {
    const optionsDebug = { collection: 'cmaups', connection };
    debug(e, optionsDebug);
  }
}
