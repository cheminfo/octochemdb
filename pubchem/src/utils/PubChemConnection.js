import delay from 'delay';
import { MongoClient } from 'mongodb';

import Debug from './Debug.js';

const debug = Debug('PubChemConnection');

export function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getLogs = async function getCollection(
  collectionName,
) {
  const logsCollection = await this.getLogsCollection();
  const _id = `${collectionName}`;
  let logs = await logsCollection.find({ _id }).next();
  if (logs === null) {
    let sources = [];
    logs = {
      _id,
      sources,
      sourcesHash: md5(JSON.stringify(sources)),
      dataStart: 0,
      dataEnd: 0,
      startSequenceID: 123,
      endSeqneceID: 123,
      status: 'updating',
    };
    await logsCollection.insertOne(logs);
  }
  return logs;
};
PubChemConnection.prototype.setLogs = async function setLogs(logs) {
  const collection = await this.getLogsCollection();
  await collection.replaceOne({ _id: logs._id }, logs);
};
PubChemConnection.prototype.getLogsCollection = async function getCollection() {
  return (await this.getDatabase()).collection('logs');
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function setProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
