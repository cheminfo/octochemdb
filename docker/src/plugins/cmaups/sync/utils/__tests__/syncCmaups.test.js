import { PubChemConnection } from '../../../../../utils/PubChemConnection.js';
import { sync } from '../../syncCmaups';

//remove process.env.TEST from parseBioactivities to use this test

test('syncCmaups', async () => {
  const connection = new PubChemConnection();
  await sync(connection);
  const collection = await connection.getCollection('cmaups');
  const collectionEntry = await collection.find({ _id: 'NPC146355' }).limit(1);
  const result = await collectionEntry.next();
  expect(result).toMatchSnapshot();
  if (connection) {
    await connection.close();
  }
});
