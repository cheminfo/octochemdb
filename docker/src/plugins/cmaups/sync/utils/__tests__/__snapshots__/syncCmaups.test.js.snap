// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`syncCmaups 1`] = `
{
  "_id": "NPC146355",
  "data": {
    "activities": [
      {
        "assay": "IC50 : 5900 nM",
        "externalRef": "PMID : 16413779",
      },
    ],
    "cid": "24992964",
    "ocl": {
      "idCode": "fa{@p@@F\\\\fo\\\\dTRbRRRafRtIJRMkqVuTEATAAEPDaBRbe@@",
      "noStereoTautomerID": "fa{@p@@F\\\\fo\\\\dTRbRRRafRtIJRMkqVuUUUUUUUU@AkuFLXIbfFXy\`VIXU\`vKX]cv@xcaNLxKbnFx{b^ExRFN\\\\YL_CGpp",
    },
    "taxonomies": [
      {
        "class": "Candidatus Borrarchaeia",
        "dbRef": {
          "$id": "NPC146355",
          "$ref": "cmaups",
        },
        "family": "Candidatus Borrarchaeaceae",
        "genus": "Candidatus Borrarchaeum",
        "order": "Candidatus Borrarchaeales",
        "phylum": "Candidatus Borrarchaeota",
        "superkingdom": "Archaea",
      },
    ],
  },
}
`;
